service/matching/task_queue_partition_manager.go

```diff
@@ -1075,3 +1086,41 @@ func (pm *taskQueuePartitionManagerImpl) getPerTypeUserData() (*persistencespb.T
 	perType := userData.GetData().GetPerType()[int32(pm.Partition().TaskType())]
 	return perType, userDataChanged, nil
 }
+
+// makePollerScalingDecision makes a decision on whether to scale pollers up or down based on the current state of the
+// task queue and the task about to be returned. Does not modify inputs.
+func (pm *taskQueuePartitionManagerImpl) makePollerScalingDecision(
+	stats *taskqueuepb.TaskQueueStats, task *internalTask, pollStartTime time.Time,
+) *sdkpb.PollerScalingDecision {
+	pd := &sdkpb.PollerScalingDecision{}
+	pollWaitTime := pm.engine.timeSource.Since(pollStartTime)
+	if stats.ApproximateBacklogCount > pm.config.PollerScalingMinimumBacklog() {
+		// Always increase when there is a backlog, even if we're a partition. It's also important to increase for
+		// sticky queues.
+		pd.PollerDelta = 1
+	} else if !pm.partition.IsRoot() {
+		// Non-root partitions don't have an appropriate view of the data to make decisions beyond backlog.
+		pd = nil
+	} else if task.source == serverenumspb.TASK_SOURCE_HISTORY &&
+		pollWaitTime >= pm.config.PollerScalingSyncMatchWaitTime() {
+		// Decrease if any poll matched after sitting idle for some configured period
+		pd.PollerDelta = -1
+	} else if stats.TasksAddRate/stats.TasksDispatchRate > pm.config.PollerScalingDispatchUpFraction() {
+		// Increase if we're adding tasks faster than we're dispatching them. This case is particularly useful when
+		// a new burst of traffic arrives. The backlog may stay at or bounce off of zero as tasks are delivered, and
+		// this allows the pollers to start scaling without accumulating a backlog.
```

Comment:
I tried for 5 minutes and could not come with an example that this is possible. i.e. have a zero backlog and stats.TasksAddRate/stats.TasksDispatchRate being >1. (Note that both rates are based on the last 30s)
Am I missing something?

Reply:
It's definitely possible because I've seen it happen (and not infrequently). I agree it's not intuitive why it happens, though. Perhaps something is not exactly accurate in the way the rates are calculated, but it seems like it happens when a bunch of tasks are added after a period of inactivity.

Reply:
The add/dispatch rates include sync match, so it's possible that a bunch of new tasks all get sync matched to waiting pollers with no backlog. That would exactly match Spencer's description.

Reply:
Ahh, that makes sense then that's definitely what's gone on.